ssl/ssl.go

```diff
@@ -265,18 +321,80 @@ func createCert(certInfo map[string]CertInfo, namespace, clusterDomain string, s
 		return nil, fmt.Errorf("Failed to sign cert: %s", err)
 	}
 
-	if len(info.PrivateKeyName) == 0 {
-		return nil, fmt.Errorf("Certificate %s created with empty private key name", id)
+	return &info, nil
+}
+func createKubeCSR(csri util.CertificateSigningRequestInterface, request []byte, info CertInfo, namespace, id string, autoApproval bool) (*CertInfo, error) {
+	csr := &certificates.CertificateSigningRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: util.ConvertNameToKey(fmt.Sprintf("%s-%s", namespace, id)),
+		},
+		Spec: certificates.CertificateSigningRequestSpec{
+			Request: request,
+			Usages: []certificates.KeyUsage{
+				certificates.UsageClientAuth,
+				certificates.UsageServerAuth,
+			},
+			// expiration is determined by apiserver and cannot be configured
+		},
 	}
-	if len(info.PrivateKey) == 0 {
-		return nil, fmt.Errorf("Certificate %s created with empty private key", id)
+	glog.Printf("create kube csr %s", csr.Name)
+	csr, err := csri.Create(csr)
+	if err == nil {
+		info.CSRName = csr.Name
 	}
-	if len(info.CertificateName) == 0 {
-		return nil, fmt.Errorf("Certificate %s created with empty certificate name", id)
+	return &info, err
+}
+
+func approveKubeCSR(csri util.CertificateSigningRequestInterface, csrName string) {
+	glog.Printf("attempting to auto-approve kube csr %s", csrName)
+
+	csr, err := csri.Get(csrName, metav1.GetOptions{})
+	if err != nil {
+		glog.Printf("unexpected error during get kube csr %s: %v", csrName, err)
+		return
 	}
-	if len(info.Certificate) == 0 {
-		return nil, fmt.Errorf("Certificate %s created with empty certificate", id)
+
+	csr.Status.Conditions = append(csr.Status.Conditions, certificates.CertificateSigningRequestCondition{
+		Type:    certificates.CertificateApproved,
+		Reason:  "Autoapproved",
+		Message: "This csr was approved automatically by scf-secrets-generator",
+	})
+	csr, err = csri.UpdateApproval(csr)
+	if err != nil {
+		glog.Printf("cannot auto-approve kube csr %s: %v\nwaiting for manual approval", csr.Name, err)
+	} else {
+		glog.Printf("kube csr %s has been auto-approved", csr.Name)
 	}
+}
 
+func waitForKubeCSR(csri util.CertificateSigningRequestInterface, info CertInfo) (*CertInfo, error) {
+	name := info.CSRName
+	var csr *certificates.CertificateSigningRequest
+	var err error
+	for retry := 0; ; retry++ {
+		csr, err = csri.Get(name, metav1.GetOptions{})
+		if err != nil {
+			return &info, fmt.Errorf("fetching kube csr %s returned error %s", name, err)
+		}
+		if len(csr.Status.Conditions) > 0 && len(csr.Status.Conditions[0].Type) > 0 {
+			break
+		}
+		switch {
+		case retry == 0:
+			glog.Printf("waiting for kube csr %s to be approved", name)
+			time.Sleep(2 * time.Second)
+		case retry < 10:
+			time.Sleep(5 * time.Second)
+		default:
+			time.Sleep(20 * time.Second)
+		}
+	}
+	cond := csr.Status.Conditions[0]
+	if cond.Type != certificates.CertificateApproved {
```

Comment:
The only possible values here are approved or denied (there are no unrelated conditions in this array), right? We can't have the approval condition later in the condition list?

Reply:
Technically it would be possible to have multiple `Conditions` in the slice, but I don't see how that makes sense.

I guess it really depends on the cert approver (you could have a custom implementation), but given that there are only 2 defined type values, I don't know what we should do. Loop through all conditions and stop at the first approved or denied type?

Reply:
That was my thought, yes.

Reply:
Fixed in c605f62